Windows install fails on dex deployment

* epinio (built from source) @ c7afb19

* Windows 10 Home, Kubernetes via Rancher Desktop

* Running `epinio install` in PowerShell

Installation fails with
```
panic: Deployment dex failed with error: Failed to extract embedded file:  - file does not exist


goroutine 52 [running]:
github.com/epinio/epinio/internal/cli/clients.(*InstallClient).Install.func1(0xc000254150, 0x24ca9b0, 0xc000034050, 0x24d5670, 0xc00044b400, 0x24d9df8, 0xc00020d0b0, 0xc00020d094)
        /abuild/projects/rancher/epinio/epinio/internal/cli/clients/install_client.go:168 +0x23f
created by github.com/epinio/epinio/internal/cli/clients.(*InstallClient).Install
        /abuild/projects/rancher/epinio/epinio/internal/cli/clients/install_client.go:165 +0x10e5
```

--- a/Makefile
+++ b/Makefile
@@ -43,6 +43,9 @@ compress:
 test: embed_files
 	ginkgo -r -p -race -failOnPending helpers internal
 
+tag:
+	@git describe --tags --abbrev=0
+
 # acceptance is not part of the unit tests, and has its own target, see below.
 
 GINKGO_NODES ?= 2

--- a/Makefile
+++ b/Makefile
@@ -7,6 +7,8 @@ export LDFLAGS += -X github.com/epinio/epinio/internal/version.Version=$(VERSION
 
 build: embed_files build-amd64
 
+build-win: embed_files build-windows
+
 build-all: embed_files build-amd64 build-arm64 build-arm32 build-windows build-darwin
 
 build-all-small:

--- a/docs/developer/howtos/development.md
+++ b/docs/developer/howtos/development.md
@@ -72,11 +72,11 @@ failing server component does not matter.
 
 If the cluster is not running on linux-amd64 it will be necessary to set
 `EPINIO_BINARY_PATH` to the correct binary to place into the epinio server
-([See here](https://github.com/epinio/epinio/blob/2c3c93f79b1019fe7895273b94f40b725ede2996/scripts/patch-epinio-deployment.sh#L19)).
+([See here](https://github.com/epinio/epinio/blob/a4b679af88d58177cecf4a5717c8c96f382058ed/scripts/patch-epinio-deployment.sh#L19)).
 
 If the client operation is performed outside of a git checkout it will be
 necessary to set `EPINIO_BINARY_TAG` to the correct tag
-([See here](https://github.com/epinio/epinio/blob/2c3c93f79b1019fe7895273b94f40b725ede2996/scripts/patch-epinio-deployment.sh#L20)).
+([See here](https://github.com/epinio/epinio/blob/a4b679af88d58177cecf4a5717c8c96f382058ed/scripts/patch-epinio-deployment.sh#L20)).
 
 The make target `tag` can be used in the checkout the binary came from to
 determine this value.

--- a/docs/developer/howtos/development.md
+++ b/docs/developer/howtos/development.md
@@ -33,7 +33,7 @@ make build
 ```
 
 This is building Epinio for linux on amd64 architecture. If you are on a
-different OS or architecture you can use one of the available `build-*` targets.
+different OS or architecture you can use one of other the available `build-*` targets.
 Look at the Makefile at the root of the project to see what is available.
 
 ### Installing Epinio
@@ -50,8 +50,6 @@ make install
 In case you're curious why `make install` is used here instead of
 `epinio install`, [look behind the curtains](behind-the-curtains.md)
 explains the details of running an Epinio dev environment.
-
-
 
 After making changes to the binary simply invoking `make
 patch-epinio-deployment` again will upload the changes into the
@@ -72,9 +70,16 @@ API credentials and certs into the client configuration file. As that
 command talks directly to the cluster and not the epinio API the
 failing server component does not matter.
 
-If the cluster is not running on linux-x64 it may be necessary to set
-`EPINIO_BINARY_PATH` to the right binary
+If the cluster is not running on linux-amd64 it will be necessary to set
+`EPINIO_BINARY_PATH` to the correct binary to place into the epinio server
 ([See here](https://github.com/epinio/epinio/blob/2c3c93f79b1019fe7895273b94f40b725ede2996/scripts/patch-epinio-deployment.sh#L19)).
+
+If the client operation is performed outside of a git checkout it will be
+necessary to set `EPINIO_BINARY_TAG` to the correct tag
+([See here](https://github.com/epinio/epinio/blob/2c3c93f79b1019fe7895273b94f40b725ede2996/scripts/patch-epinio-deployment.sh#L20)).
+
+The make target `tag` can be used in the checkout the binary came from to
+determine this value.
 
 Also, the default `make build` target builds a dynamically linked
 binary. This can cause issues if for example the glibc library in the
@@ -85,3 +90,21 @@ command like:
 ```
 GOARCH="amd64" GOOS="linux" CGO_ENABLED=0 go build -o dist/epinio-linux-amd64
 ```
+
+#### Mixed Windows/Linux Scenario
+
+A concrete example of the above would be the installation of Epinio from a
+Windows host without a checkout, to a Linux-based cluster.
+
+In that scenario the Windows host has to have both windows-amd64 and linux-amd64
+binaries. The first to perform the installation, the second for
+`EPINIO_BINARY_PATH` to be put into the server.
+
+Furthermore, as the Windows host is without a checkout, the tag has to be
+determined in the actual checkout and set into `EPINIO_BINARY_PATH`.
+
+Lastly, do not forget to set up a proper domain so that the client can talk to
+the server after installation is done. While during installation only a suitable
+`KUBECONFIG` is required after the client will go and use the information from
+the ingress, and that then has to properly resolve in the DNS.
+

--- a/go.mod
+++ b/go.mod
@@ -9,7 +9,6 @@ require (
 	github.com/avast/retry-go v3.0.0+incompatible
 	github.com/briandowns/spinner v1.12.0
 	github.com/codeskyblue/kexec v0.0.0-20180119015717-5a4bed90d99a
-	github.com/coreos/go-oidc/v3 v3.0.0
 	github.com/fatih/color v1.12.0
 	github.com/go-logr/logr v0.4.0
 	github.com/go-logr/stdr v0.4.0
@@ -32,7 +31,6 @@ require (
 	github.com/stretchr/testify v1.7.0 // indirect
 	github.com/tektoncd/pipeline v0.23.0
 	golang.org/x/crypto v0.0.0-20210322153248-0c34fe9e7dc2
-	golang.org/x/oauth2 v0.0.0-20210126194326-f9ce19ea3013
 	golang.org/x/sys v0.0.0-20210514084401-e8d321eab015 // indirect
 	golang.org/x/tools v0.1.1 // indirect
 	k8s.io/api v0.20.5

--- a/go.mod
+++ b/go.mod
@@ -17,6 +17,7 @@ require (
 	github.com/imdario/mergo v0.3.11 // indirect
 	github.com/julienschmidt/httprouter v1.3.0
 	github.com/kyokomi/emoji v2.2.4+incompatible
+	github.com/mattn/go-isatty v0.0.12
 	github.com/maxbrunsfeld/counterfeiter/v6 v6.3.0
 	github.com/mholt/archiver/v3 v3.5.0
 	github.com/olekukonko/tablewriter v0.0.4

--- a/helpers/termui/ui.go
+++ b/helpers/termui/ui.go
@@ -194,7 +194,7 @@ func (u *Message) Msg(message string) {
 		message = color.RedString(message)
 	}
 
-	fmt.Printf("%s", message)
+	fmt.Fprintf(color.Output, "%s", message)
 
 	for _, interaction := range u.interactions {
 		switch interaction.variant {
@@ -211,17 +211,17 @@ func (u *Message) Msg(message string) {
 		case show:
 			switch interaction.valueType {
 			case tBool:
-				fmt.Printf("%s: %s\n", emoji.Sprint(interaction.name), color.MagentaString("%t", interaction.value))
+				fmt.Fprintf(color.Output, "%s: %s\n", emoji.Sprint(interaction.name), color.MagentaString("%t", interaction.value))
 			case tInt:
-				fmt.Printf("%s: %s\n", emoji.Sprint(interaction.name), color.CyanString("%d", interaction.value))
+				fmt.Fprintf(color.Output, "%s: %s\n", emoji.Sprint(interaction.name), color.CyanString("%d", interaction.value))
 			case tString:
-				fmt.Printf("%s: %s\n", emoji.Sprint(interaction.name), color.GreenString("%s", interaction.value))
+				fmt.Fprintf(color.Output, "%s: %s\n", emoji.Sprint(interaction.name), color.GreenString("%s", interaction.value))
 			}
 		}
 	}
 
 	for idx, headers := range u.tableHeaders {
-		table := tablewriter.NewWriter(os.Stdout)
+		table := tablewriter.NewWriter(color.Output)
 		table.SetHeader(headers)
 		table.SetBorders(tablewriter.Border{Left: true, Top: false, Right: true, Bottom: false})
 		table.SetCenterSeparator("|")

--- a/internal/cli/clients/install_client.go
+++ b/internal/cli/clients/install_client.go
@@ -14,6 +14,7 @@ import (
 	"github.com/epinio/epinio/helpers/kubernetes"
 	"github.com/epinio/epinio/helpers/termui"
 	"github.com/epinio/epinio/helpers/tracelog"
+	"github.com/epinio/epinio/internal/cli/config"
 	"github.com/epinio/epinio/internal/duration"
 	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
@@ -33,6 +34,15 @@ type InstallClient struct {
 }
 
 func NewInstallClient(ctx context.Context, options *kubernetes.InstallationOptions) (*InstallClient, func(), error) {
+	// We do this for the side effect: colorized output vs not.
+	// May also extend the internal CA cert pool.
+	// This and everything else done by loading does not matter.
+	// The later phases of the installer will write a new config.
+	_, err := config.Load()
+	if err != nil {
+		return nil, nil, err
+	}
+
 	cluster, err := kubernetes.GetCluster(ctx)
 	if err != nil {
 		return nil, nil, err

--- a/internal/cli/debug.go
+++ b/internal/cli/debug.go
@@ -1,0 +1,45 @@
+package cli
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/mattn/go-isatty"
+	"github.com/spf13/cobra"
+)
+
+var ()
+
+func init() {
+	CmdDebug.AddCommand(CmdDebugTTY)
+}
+
+// CmdDebug implements the epinio debug command
+var CmdDebug = &cobra.Command{
+	Hidden:        true,
+	Use:           "debug",
+	Short:         "Dev Tools",
+	Long:          `Developer Tools. Hidden From Regular User.`,
+	Args:          cobra.ExactArgs(0),
+	SilenceErrors: true,
+	SilenceUsage:  true,
+}
+
+// CmdDebug implements the epinio debug command
+var CmdDebugTTY = &cobra.Command{
+	Use:   "tty",
+	Short: "Running In a Terminal?",
+	Long:  `Running In a Terminal?`,
+	Args:  cobra.ExactArgs(0),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cmd.SilenceUsage = true
+		if isatty.IsTerminal(os.Stdout.Fd()) {
+			fmt.Println("Is Terminal")
+		} else if isatty.IsCygwinTerminal(os.Stdout.Fd()) {
+			fmt.Println("Is Cygwin/MSYS2 Terminal")
+		} else {
+			fmt.Println("Is Not Terminal")
+		}
+		return nil
+	},
+}

--- a/internal/cli/root.go
+++ b/internal/cli/root.go
@@ -89,6 +89,8 @@ func init() {
 	rootCmd.AddCommand(CmdService)
 	rootCmd.AddCommand(CmdServer)
 	rootCmd.AddCommand(cmdVersion)
+	// Hidden command providing developer tools
+	rootCmd.AddCommand(CmdDebug)
 }
 
 var cmdVersion = &cobra.Command{

--- a/scripts/patch-epinio-deployment.sh
+++ b/scripts/patch-epinio-deployment.sh
@@ -16,8 +16,14 @@
 # Patching the deployment forces the pod to restart with a now existing image
 # and the correct binary is in place.
 
-export EPINIO_BINARY_PATH=${EPINIO_BINARY_PATH:-'dist/epinio-linux-amd64'}
+export EPINIO_BINARY_PATH="${EPINIO_BINARY_PATH:-dist/epinio-linux-amd64}"
+export EPINIO_BINARY_TAG="${EPINIO_BINARY_TAG:-$(git describe --tags --abbrev=0)}"
 
+echo
+echo Configuration
+echo "  - Binary: ${EPINIO_BINARY_PATH}"
+echo "  - Tag:    ${EPINIO_BINARY_TAG}"
+echo
 echo "Creating the PVC"
 cat <<EOF | kubectl apply -f -
 ---
@@ -60,7 +66,19 @@ echo "Waiting for dummy pod to be ready"
 kubectl wait --for=condition=ready pod -n epinio epinio-copier
 
 echo "Copying the binary on the PVC"
-kubectl cp ${EPINIO_BINARY_PATH} epinio/epinio-copier:/epinio/epinio
+# Notes
+# 1. kubectl cp breaks because of the colon in `pod:path`. Thus the more complex tar construction.
+# 2. Cannot use absolute paths, i.e. `/foo`. This is a disk-relative path on Windows, and gets
+#    expanded weirdly by the kubectl commands.
+#    Relative paths are ok, as the default CWD in the container is the root (`/`).
+#    I.e. `foo` is `/foo` pod-side.
+
+( cd       "$(dirname  "${EPINIO_BINARY_PATH}")"
+  tar cf - "$(basename "${EPINIO_BINARY_PATH}")"
+) | kubectl exec -i -n epinio -c copier epinio-copier -- tar xf -
+kubectl exec -i -n epinio -c copier epinio-copier -- mv "$(basename "${EPINIO_BINARY_PATH}")" epinio/epinio
+kubectl exec -i -n epinio -c copier epinio-copier -- chmod ugo+x epinio/epinio
+kubectl exec -i -n epinio -c copier epinio-copier -- ls -l epinio
 
 echo "Patching the epinio-server deployment to use the copied binary"
 read -r -d '' PATCH <<EOF
@@ -77,7 +95,7 @@ read -r -d '' PATCH <<EOF
         ],
         "containers": [{
           "name": "epinio-server",
-          "image": "splatform/epinio-base:$(git describe --tags --abbrev=0)",
+          "image": "splatform/epinio-base:${EPINIO_BINARY_TAG}",
           "command": [
             "/epinio-binary/epinio",
             "server"

